In [14]:
#import all the packages
import pandas as pd
import numpy as np
import tweepy
import requests
import json

# Gather Data from different Sources

### 1. Load Twitter archive data from csv file

In [15]:
#Load dataset
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [16]:
#Display the dataset
twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

### 2.Load  Image prediction data using url

In [17]:
#Gather Image Predictions File from a url online
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
response

<Response [200]>

In [18]:
open('image_predictions.tsv', 'wb').write(response.content)

335079

In [19]:
#Create dataframe from TSV file
twitter_image_predictions= pd.read_csv("image_predictions.tsv",sep='\t')
twitter_image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

### Load Retweet/Favourite count data using Twitter API

In [14]:
# Download data using Twitter API
consumer_key = 'MY CONSUMER KEY'
consumer_secret = 'MY CONSUMER SECRET'
access_token = 'MY ACCESS TOKEN'
access_secret = 'MY ACCESS SECRET'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth,wait_on_rate_limit = True, wait_on_rate_limit_notify = True)


In [18]:
#Store data in tweet_json.txt
tweet_ids = df.tweet_id.values

with open('tweet_json.txt', 'w') as output:
    for tweet_id in tweet_ids:
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            json.dump(tweet._json, output)
            output.write('\n')
        except tweepy.TweepError:
            print('Error: tweet missing')

Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing
Error: tweet missing


In [20]:
#Converting relevant JSON data to dataframe
tweets_data=[]
with open('tweet_json.txt','r') as json_file:
    line = json_file.readline()
    while line:
        data=json.loads(line)
        
        tweet_id= data['id']
        retweet_counts=data['retweet_count']
        favorite_counts=data['favorite_count']
        
        json_data={'tweet_id':tweet_id, 'retweet_count':retweet_counts,'favorite_count':favorite_counts}
        
        tweets_data.append(json_data)
        
        line = json_file.readline()
        
df_tweets_data=pd.DataFrame(tweets_data,columns=['tweet_id',
                                              'retweet_count',
                                              'favorite_count'])
df_tweets_data.head(5)


tweet_id  retweet_count  favorite_count
0  892420643555336193           7717           36249
1  892177421306343426           5703           31257
2  891815181378084864           3781           23533
3  891689557279858688           7870           39526
4  891327558926688256           8486           37740

# Access Data

## 1. Twitter Archived Enhanced Data

In [21]:
#display data
twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [22]:
#Get complete info about the columns
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [23]:
#Get descriptive statistics
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [24]:
#display value count of source field
twitter_archive.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

In [25]:
#Check duplicate records
sum(twitter_archive.duplicated())

0

In [26]:
#Check value count of each name
twitter_archive.name.value_counts()

None        745
a            55
Charlie      12
Cooper       11
Oliver       11
           ... 
Winifred      1
Nugget        1
Brat          1
Jeb           1
Jeffri        1
Name: name, Length: 957, dtype: int64

In [27]:
#Check null values in twitter_archive dataset  
twitter_archive.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

In [28]:
#Check value count of  puppo

twitter_archive.puppo.value_counts()

None     2326
puppo      30
Name: puppo, dtype: int64

In [29]:
#Check value count of  pupper
twitter_archive.pupper.value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [30]:
#Check value count of  floofer
twitter_archive.floofer.value_counts()

None       2346
floofer      10
Name: floofer, dtype: int64

In [31]:
#Check value count of  pupper
twitter_archive.pupper.value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

## Image prediction Data

In [32]:
#Display few rows of twitter_image_predictions dataset
twitter_image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [33]:
# View information of twitter_image_predictions DataFrame
twitter_image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [34]:
#get descriptive statistics
twitter_image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [35]:
#Display no of rows and columns
twitter_image_predictions.shape 

(2075, 12)

In [36]:
#Check value count of each p1

twitter_image_predictions.p1.value_counts()

golden_retriever      150
Labrador_retriever    100
Pembroke               89
Chihuahua              83
pug                    57
                     ... 
sandbar                 1
bighorn                 1
alp                     1
conch                   1
groenendael             1
Name: p1, Length: 378, dtype: int64

In [37]:
#Check value count of each p2
twitter_image_predictions.p2.value_counts()

Labrador_retriever    104
golden_retriever       92
Cardigan               73
Chihuahua              44
Pomeranian             42
                     ... 
timber_wolf             1
goose                   1
dam                     1
sandal                  1
triceratops             1
Name: p2, Length: 405, dtype: int64

In [38]:
#Check value count of each p3
twitter_image_predictions.p3.value_counts()

Labrador_retriever    79
Chihuahua             58
golden_retriever      48
Eskimo_dog            38
kelpie                35
                      ..
hammerhead             1
moped                  1
wolf_spider            1
ibex                   1
buckeye                1
Name: p3, Length: 408, dtype: int64

In [39]:
#Check for duplicates rows
sum(df_tweets_data.duplicated())

0

## 3.Twitter API data


In [40]:
#Display few rows of twitter_image_predictions dataset

df_tweets_data.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           7717           36249
1  892177421306343426           5703           31257
2  891815181378084864           3781           23533
3  891689557279858688           7870           39526
4  891327558926688256           8486           37740

In [41]:
# View information of df_tweets_data DataFrame

df_tweets_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
tweet_id          2331 non-null int64
retweet_count     2331 non-null int64
favorite_count    2331 non-null int64
dtypes: int64(3)
memory usage: 54.8 KB


In [42]:
#get descriptive statistics

df_tweets_data.describe()

tweet_id  retweet_count  favorite_count
count  2.331000e+03    2331.000000     2331.000000
mean   7.419079e+17    2706.800515     7568.296868
std    6.823170e+16    4577.088757    11745.064808
min    6.660209e+17       1.000000        0.000000
25%    6.782670e+17     548.000000     1319.500000
50%    7.182469e+17    1269.000000     3293.000000
75%    7.986692e+17    3145.000000     9264.000000
max    8.924206e+17   77894.000000   156324.000000

In [43]:
#Check duplicate records
sum(df_tweets_data.duplicated())

0

## Quality Issues:
##### 1. As there are 78 records for  " in_reply_to_status_id"   and  "in_reply_to_user_id"  columns which is very less in count , so both columns are not helpful in analysis, that's why drop these columns.
#### 2. All columns related to 'retweets' contains only 181 records ,so they are not helpful in analysis, that's why drop these columns.
#### 3. Change datatype of timestamp from object to datetime
#### 4. Extract the string between  from Source field between a href tag.
#### 5. Missing value in expanded_urls columns
#### 6. Exclude  zero values from the rating_denuminator.
#### 7. Replace the value 'None' with the NaN (missing value), so that it become easy to find missing values directly by using built-in functions
#### 8. Missing value in name field..
#### 9. Change the column names of image prediction table from p1 ,p2,p3 to prediction_1,prediction_1,prediction_1 respectively.


## Tidiness Issues:
#### 1. change last  4 columns  'doggo', 'floofer','pupper', 'puppo' into one column as dog_stage
#### 2. Merge all the tables to make only one master table.




## Cleaning Data

In [44]:
#Take copies of original DataFrames 
twitter_archive_clean=twitter_archive.copy()
twitter_image_predictions_clean=twitter_image_predictions.copy()
df_tweets_data_clean=df_tweets_data.copy()

### Resolve all  Quality issue

#### Define

##### 1. As there are 78 records for  " in_reply_to_status_id"   and  "in_reply_to_user_id"  columns which is very less in count , so both columns are not helpful in analysis, that's why drop these columns.

#### Code

In [45]:
#Drop both reply columns
twitter_archive_clean.drop(['in_reply_to_status_id','in_reply_to_user_id'],inplace=True,axis=1) 

#### Test

In [46]:
#Check columns list
twitter_archive_clean.columns

Index(['tweet_id', 'timestamp', 'source', 'text', 'retweeted_status_id',
       'retweeted_status_user_id', 'retweeted_status_timestamp',
       'expanded_urls', 'rating_numerator', 'rating_denominator', 'name',
       'doggo', 'floofer', 'pupper', 'puppo'],
      dtype='object')

#### Define

#### 2. All columns related to 'retweets' contains only 181 records ,so they are not helpful in analysis, that's why drop these columns.



#### Code

In [47]:
#Drop all 3 retweet columns
twitter_archive_clean.drop(['retweeted_status_id','retweeted_status_user_id','retweeted_status_timestamp'],axis = 1,inplace= True)

#### Test

In [48]:
#Check columns list

twitter_archive_clean.columns

Index(['tweet_id', 'timestamp', 'source', 'text', 'expanded_urls',
       'rating_numerator', 'rating_denominator', 'name', 'doggo', 'floofer',
       'pupper', 'puppo'],
      dtype='object')

#### Define

#### 3. Change datatype of timestamp from object to datetime

#### Code

In [49]:
#change datatype  of timestamp to datetime 
twitter_archive_clean['timestamp'] = pd.to_datetime(twitter_archive_clean['timestamp'])

#### Test

In [50]:
#Check datatpe of timestamp(changed to datetime)
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2356 non-null int64
timestamp             2356 non-null datetime64[ns, UTC]
source                2356 non-null object
text                  2356 non-null object
expanded_urls         2297 non-null object
rating_numerator      2356 non-null int64
rating_denominator    2356 non-null int64
name                  2356 non-null object
doggo                 2356 non-null object
floofer               2356 non-null object
pupper                2356 non-null object
puppo                 2356 non-null object
dtypes: datetime64[ns, UTC](1), int64(3), object(8)
memory usage: 221.0+ KB


#### Define

#### 4. Extract the string between  from Source field between a href tag 

#### Code

In [51]:
#Extract the string from source field
twitter_archive_clean['source'] = twitter_archive_clean['source'].str.extract('^<a.+>(.+)</a>$')


#### Test

In [52]:
twitter_archive_clean['source'].value_counts()

Twitter for iPhone     2221
Vine - Make a Scene      91
Twitter Web Client       33
TweetDeck                11
Name: source, dtype: int64

#### Define

#### 5. Missing value in expanded_urls columns

#### Code

In [53]:
#Drop the missing  values from expanded_urls
twitter_archive_clean['expanded_urls'].dropna(inplace= True)

#### Test

In [54]:
#No null value
sum(twitter_archive_clean.expanded_urls.isnull())

0

#### Define

#### 6. Exclude  zero values from the rating_denuminator


#### Code

In [55]:
#Check value count of each rating_denominator
twitter_archive_clean.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

#### It contains 0 for one record

In [56]:
#Get the index of the row
twitter_archive_clean[twitter_archive_clean.rating_denominator==0] 

tweet_id                 timestamp              source  \
313  835246439529840640 2017-02-24 21:54:03+00:00  Twitter for iPhone   

                                                  text expanded_urls  \
313  @jonnysun @Lin_Manuel ok jomny I know you're e...           NaN   

     rating_numerator  rating_denominator  name doggo floofer pupper puppo  
313               960                   0  None  None    None   None  None

In [57]:
#drop inde 313 that contains 0 in rating_denominator
twitter_archive_clean.drop(index=313,inplace= True) 

#### Test

In [58]:
#Result: No record found with 0 rating_denominator
twitter_archive_clean[twitter_archive_clean.rating_denominator==0] 

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

#### Define


#### 7. Replace the value 'None' with the NaN (missing value), so that it become easy to find missing values directly by using built-in functions




#### Code

In [59]:
#Replace all None with Nan
twitter_archive_clean =  twitter_archive_clean.replace('None', np.nan)

#### Test

In [60]:
#Result:Now we can see  which column contains null values 
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2355 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2355 non-null int64
timestamp             2355 non-null datetime64[ns, UTC]
source                2355 non-null object
text                  2355 non-null object
expanded_urls         2297 non-null object
rating_numerator      2355 non-null int64
rating_denominator    2355 non-null int64
name                  1611 non-null object
doggo                 97 non-null object
floofer               10 non-null object
pupper                257 non-null object
puppo                 30 non-null object
dtypes: datetime64[ns, UTC](1), int64(3), object(8)
memory usage: 239.2+ KB


#### Define

#### 8. Missing value in name field.

#### Code

In [61]:
#Check count of null in name field
twitter_archive_clean.name.isnull().sum()

744

In [62]:
#Drop all null values
twitter_archive_clean.name.dropna(inplace=True)

#### Test

In [63]:
#Result: No null value found
twitter_archive_clean.name.isnull().sum()

0

#### Define 

#### 9. Change the column names of image prediction table from p1 ,p2,p3 to prediction_1,prediction_1,prediction_1 respectively

#### Code

In [64]:
#Rename the required columns
twitter_image_predictions_clean.rename(columns={ 'p1':'prediction_1','p2':'prediction_2','p3':'prediction_3'},inplace=True)

#### Test

In [65]:
#Result: All columns name got changed 
twitter_image_predictions_clean.columns

Index(['tweet_id', 'jpg_url', 'img_num', 'prediction_1', 'p1_conf', 'p1_dog',
       'prediction_2', 'p2_conf', 'p2_dog', 'prediction_3', 'p3_conf',
       'p3_dog'],
      dtype='object')

### Resolve tidiness issue

#### Define

#### 10. change last  4 columns  'doggo', 'floofer','pupper', 'puppo' into one column

#### Code

In [66]:
#changing last  4 columns  'doggo', 'floofer','pupper', 'puppo' into one column i.e. dog_stage
Stages_list = ['doggo','pupper','floofer','puppo']
Remaining_columns = [i for i in twitter_archive_clean.columns.tolist() if i not in Stages_list]


twitter_archive_clean = pd.melt(twitter_archive_clean, id_vars = Remaining_columns, value_vars = Stages_list, var_name = 'stages', value_name = 'dog_stage')


In [67]:
# Delete column 'stages'
twitter_archive_clean = twitter_archive_clean.drop('stages',axis= 1)

In [68]:
#result: New Column Added i.e. dog_stage which is the replacemnet of 'doggo', 'floofer','pupper', 'puppo' columns
twitter_archive_clean.head(50)

tweet_id                 timestamp              source  \
0   892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1   892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2   891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3   891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4   891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   
5   891087950875897856 2017-07-29 00:08:17+00:00  Twitter for iPhone   
6   890971913173991426 2017-07-28 16:27:12+00:00  Twitter for iPhone   
7   890729181411237888 2017-07-28 00:22:40+00:00  Twitter for iPhone   
8   890609185150312448 2017-07-27 16:25:51+00:00  Twitter for iPhone   
9   890240255349198849 2017-07-26 15:59:51+00:00  Twitter for iPhone   
10  890006608113172480 2017-07-26 00:31:25+00:00  Twitter for iPhone   
11  889880896479866881 2017-07-25 16:11:53+00:00  Twitter for iPhone   
12  889665388333682689 2017-07-25 01:55:32+00:00  Twitter for iPhone   
13  889638837579907072 2017-07-25 00:10:02+00:00  Twitter for iPhone   
14  889531135344209921 2017-07-24 17:02:04+00:00  Twitter for iPhone   
15  889278841981685760 2017-07-24 00:19:32+00:00  Twitter for iPhone   
16  888917238123831296 2017-07-23 00:22:39+00:00  Twitter for iPhone   
17  888804989199671297 2017-07-22 16:56:37+00:00  Twitter for iPhone   
18  888554962724278272 2017-07-22 00:23:06+00:00  Twitter for iPhone   
19  888202515573088257 2017-07-21 01:02:36+00:00  Twitter for iPhone   
20  888078434458587136 2017-07-20 16:49:33+00:00  Twitter for iPhone   
21  887705289381826560 2017-07-19 16:06:48+00:00  Twitter for iPhone   
22  887517139158093824 2017-07-19 03:39:09+00:00  Twitter for iPhone   
23  887473957103951883 2017-07-19 00:47:34+00:00  Twitter for iPhone   
24  887343217045368832 2017-07-18 16:08:03+00:00  Twitter for iPhone   
25  887101392804085760 2017-07-18 00:07:08+00:00  Twitter for iPhone   
26  886983233522544640 2017-07-17 16:17:36+00:00  Twitter for iPhone   
27  886736880519319552 2017-07-16 23:58:41+00:00  Twitter for iPhone   
28  886680336477933568 2017-07-16 20:14:00+00:00  Twitter for iPhone   
29  886366144734445568 2017-07-15 23:25:31+00:00  Twitter for iPhone   
30  886267009285017600 2017-07-15 16:51:35+00:00  Twitter for iPhone   
31  886258384151887873 2017-07-15 16:17:19+00:00  Twitter for iPhone   
32  886054160059072513 2017-07-15 02:45:48+00:00  Twitter for iPhone   
33  885984800019947520 2017-07-14 22:10:11+00:00  Twitter for iPhone   
34  885528943205470208 2017-07-13 15:58:47+00:00  Twitter for iPhone   
35  885518971528720385 2017-07-13 15:19:09+00:00  Twitter for iPhone   
36  885311592912609280 2017-07-13 01:35:06+00:00  Twitter for iPhone   
37  885167619883638784 2017-07-12 16:03:00+00:00  Twitter for iPhone   
38  884925521741709313 2017-07-12 00:01:00+00:00  Twitter for iPhone   
39  884876753390489601 2017-07-11 20:47:12+00:00  Twitter for iPhone   
40  884562892145688576 2017-07-11 00:00:02+00:00  Twitter for iPhone   
41  884441805382717440 2017-07-10 15:58:53+00:00  Twitter for iPhone   
42  884247878851493888 2017-07-10 03:08:17+00:00  Twitter for iPhone   
43  884162670584377345 2017-07-09 21:29:42+00:00  Twitter for iPhone   
44  883838122936631299 2017-07-09 00:00:04+00:00  Twitter for iPhone   
45  883482846933004288 2017-07-08 00:28:19+00:00  Twitter for iPhone   
46  883360690899218434 2017-07-07 16:22:55+00:00  Twitter for iPhone   
47  883117836046086144 2017-07-07 00:17:54+00:00  Twitter for iPhone   
48  882992080364220416 2017-07-06 15:58:11+00:00  Twitter for iPhone   
49  882762694511734784 2017-07-06 00:46:41+00:00  Twitter for iPhone   

                                                 text  \
0   This is Phineas. He's a mystical boy. Only eve...   
1   This is Tilly. She's just checking pup on you....   
2   This is Archie. He is a rare Norwegian Pouncin...   
3   This is Darla. She commenced a snooze mid meal...   
4   This is Franklin. He would like you to stop ca...   

#### Resolving the Quality issue after adding new column(dog_stage)

In [69]:
#Delete dog_stage duplicates
twitter_archive_clean = twitter_archive_clean.sort_values('dog_stage').drop_duplicates('tweet_id', keep = 'first')

In [70]:
twitter_archive_clean.dog_stage.value_counts()

pupper     245
doggo       97
puppo       29
floofer      9
Name: dog_stage, dtype: int64

#### Define

#### 11. Merge all the tables to make only one master table

#### Code

In [71]:
#Make master_table dataframe that contains data from all the three tables
master_table = pd.merge(twitter_archive_clean,twitter_image_predictions_clean,on='tweet_id',how='inner')

In [72]:
master_table=master_table.merge(df_tweets_data_clean,on='tweet_id',how='inner')

#### Test

In [73]:
#Result: master_table contains all the columns data from 3 tables
master_table.head()

tweet_id                 timestamp              source  \
0  890240255349198849 2017-07-26 15:59:51+00:00  Twitter for iPhone   
1  759793422261743616 2016-07-31 16:50:42+00:00  Twitter for iPhone   
2  760893934457552897 2016-08-03 17:43:45+00:00  Twitter for iPhone   
3  767191397493538821 2016-08-21 02:47:37+00:00  Twitter for iPhone   
4  768193404517830656 2016-08-23 21:09:14+00:00  Twitter for iPhone   

                                                text  \
0  This is Cassie. She is a college pup. Studying...   
1  Meet Maggie &amp; Lila. Maggie is the doggo, L...   
2  This is Wishes. He has the day off. Daily stru...   
3  I don't know any of the backstory behind this ...   
4  Meet Gerald. He's a fairly exotic doggo. Floof...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/890240255...                14   
1  https://twitter.com/dog_rates/status/759793422...                12   
2  https://twitter.com/dog_rates/status/760893934...                11   
3  https://twitter.com/dog_rates/status/767191397...                13   
4  https://twitter.com/dog_rates/status/768193404...                 8   

   rating_denominator    name dog_stage  \
0                  10  Cassie     doggo   
1                  10  Maggie     doggo   
2                  10  Wishes     doggo   
3                  10     NaN     doggo   
4                  10  Gerald     doggo   

                                           jpg_url  ...   p1_conf p1_dog  \
0  https://pbs.twimg.com/media/DFrEyVuW0AAO3t9.jpg  ...  0.511319   True   
1  https://pbs.twimg.com/media/CotUFZEWcAA2Pku.jpg  ...  0.985876   True   
2  https://pbs.twimg.com/media/Co88_ujWEAErCg7.jpg  ...  0.113992   True   
3  https://pbs.twimg.com/media/CqWcgcqWcAI43jm.jpg  ...  0.708665  False   
4  https://pbs.twimg.com/media/Cqkr0wiW8AAn2Oi.jpg  ...  0.396984  False   

         prediction_2   p2_conf p2_dog  prediction_3   p3_conf p3_dog  \
0            Cardigan  0.451038   True     Chihuahua  0.029248   True   
1  Labrador_retriever  0.001948   True        kuvasz  0.001752   True   
2      cocker_spaniel  0.105780   True        borzoi  0.073935   True   
3           boathouse  0.110056  False          pier  0.039532  False   
4                 ram  0.300851  False       cheetah  0.094474  False   

   retweet_count  favorite_count  
0           6703           29884  
1           1933            6009  
2            986            3847  
3           3809           12408  
4           3584           10979  

[5 rows x 22 columns]

In [75]:
#Saving the dataframe to csv file
master_table.to_csv('twitter_archive_master.csv',index=False)


### Note: Data visualisation and Insights are include in act_report.ipynb notebook.